homeassistant/components/nmap_tracker/__init__.py

```diff
@@ -1 +1,372 @@
+                self.devices.tracked[entry.unique_id] = NmapDevice(
+                    entry.unique_id,
+                    None,
+                    entry.original_name,
+                    None,
+                    self._async_get_vendor(entry.unique_id),
+                    "Device not found in initial scan",
+                    now,
+                    1,
+                )
+                async_dispatcher_send(
+                    self._hass, self.signal_device_missing, entry.unique_id
+                )
+
+    def _run_nmap_scan(self):
+        """Run nmap and return the result."""
+        options = self._build_options()
+        if not self._scanner:
+            self._scanner = PortScanner()
+        _LOGGER.debug("Scanning %s with args: %s", self._hosts, options)
+        for attempt in range(MAX_SCAN_ATTEMPTS):
+            try:
+                result = self._scanner.scan(
+                    hosts=" ".join(self._hosts),
+                    arguments=options,
+                    timeout=TRACKER_SCAN_INTERVAL * 10,
+                )
+                break
+            except PortScannerError as ex:
+                if attempt < (MAX_SCAN_ATTEMPTS - 1) and NMAP_TRANSIENT_FAILURE in str(
+                    ex
+                ):
+                    _LOGGER.debug("Nmap saw transient error %s", NMAP_TRANSIENT_FAILURE)
+                    continue
+                raise
+        _LOGGER.debug(
+            "Finished scanning %s with args: %s",
+            self._hosts,
+            options,
+        )
+        return result
+
+    @callback
+    def _async_increment_device_offline(self, ipv4, reason):
+        """Mark an IP offline."""
+        if not (formatted_mac := self.devices.ipv4_last_mac.get(ipv4)):
+            return
+        if not (device := self.devices.tracked.get(formatted_mac)):
+            # Device was unloaded
+            return
+        device.offline_scans += 1
+        if device.offline_scans < OFFLINE_SCANS_TO_MARK_UNAVAILABLE:
+            return
+        device.reason = reason
+        async_dispatcher_send(self._hass, signal_device_update(formatted_mac), False)
+        del self.devices.ipv4_last_mac[ipv4]
+
+    async def _async_run_nmap_scan(self):
+        """Scan the network for devices and dispatch events."""
+        result = await self._hass.async_add_executor_job(self._run_nmap_scan)
+        if self._stopping:
+            return
+
+        devices = self.devices
+        entry_id = self._entry_id
+        now = dt_util.now()
+        for ipv4, info in result["scan"].items():
+            status = info["status"]
+            reason = status["reason"]
+            if status["state"] != "up":
+                self._async_increment_device_offline(ipv4, reason)
+                continue
+            # Mac address only returned if nmap ran as root
+            mac = info["addresses"].get("mac") or get_mac_address(ip=ipv4)
```

Comment:
I don't believe `get_mac_address` is async safe.

Reply:
Maybe this is the reason for the slow downs that made us revert this?

Reply:
🤦 that would do it, and it makes way more sense then suddenly a problem with subprocesses.

Reply:
fixed in https://github.com/home-assistant/core/pull/54715